tests/service/TokenizerTests.fs

```diff
@@ -54,5 +54,164 @@ let ``Tokenizer test 1``() =
+           ("WHITESPACE", " "); ("EQUALS", "="); ("WHITESPACE", " ");
+           ("STRING_TEXT", "$\"\"\""); ("STRING_TEXT", "abc"); ("STRING_TEXT", " ");
+           ("INTERP_STRING_BEGIN_PART", "{"); ("INT32", "1"); ("PLUS_MINUS_OP", "+")]);
+         (1,
+          [("WHITESPACE", " "); ("INT32", "1"); ("STRING_TEXT", "}");
+           ("STRING_TEXT", " "); ("STRING_TEXT", "def"); ("INTERP_STRING_END", "\"\"\"")])]
+  
+    if actual <> expected then 
+        printfn "actual   = %A" actual
+        printfn "expected = %A" expected
+        Assert.Fail(sprintf "actual and expected did not match,actual =\n%A\nexpected=\n%A\n" actual expected)
+
+[<Test>]
+// checks nested '{' and nested single-quote strings
+let ``Tokenizer test - multi-line nested string interpolation``() =
+    let tokenizedLines = 
+      tokenizeLines
+        [| "let hello1t = $\"\"\"abc {\"a\" +               "
+           "                          {                     "
+           "                           contents = \"b\"     "
+           "                          }.contents            "
+           "                         } def\"\"\"" |]
+
+    let actual = 
+        [ for lineNo, lineToks in tokenizedLines do
+            yield lineNo, [ for str, info in lineToks do yield info.TokenName, str ] ]
+    let expected = 
+        [(0,
+          [("LET", "let"); ("WHITESPACE", " "); ("IDENT", "hello1t");
+           ("WHITESPACE", " "); ("EQUALS", "="); ("WHITESPACE", " ");
+           ("STRING_TEXT", "$\"\"\""); ("STRING_TEXT", "abc"); ("STRING_TEXT", " ");
+           ("INTERP_STRING_BEGIN_PART", "{"); ("STRING_TEXT", "\""); ("STRING_TEXT", "a");
+           ("STRING", "\""); ("WHITESPACE", " "); ("PLUS_MINUS_OP", "+");
+           ("WHITESPACE", "               ")]);
+         (1,
+          [("WHITESPACE", "                          "); ("LBRACE", "{");
+           ("WHITESPACE", "                     ")]);
+         (2,
+          [("WHITESPACE", "                           "); ("IDENT", "contents");
+           ("WHITESPACE", " "); ("EQUALS", "="); ("WHITESPACE", " ");
+           ("STRING_TEXT", "\""); ("STRING_TEXT", "b"); ("STRING", "\"");
+           ("WHITESPACE", "     ")]);
+         (3,
+          [("WHITESPACE", "                          "); ("RBRACE", "}"); ("DOT", ".");
+           ("IDENT", "contents"); ("WHITESPACE", "            ")]);
+         (4,
+          [("WHITESPACE", "                         "); ("STRING_TEXT", "}");
+           ("STRING_TEXT", " "); ("STRING_TEXT", "def"); ("INTERP_STRING_END", "\"\"\"")])]
+  
+    if actual <> expected then 
+        printfn "actual   = %A" actual
+        printfn "expected = %A" expected
+        Assert.Fail(sprintf "actual and expected did not match,actual =\n%A\nexpected=\n%A\n" actual expected)
+
+[<Test>]
+let ``Tokenizer test - single-line nested string interpolation``() =
+    let tokenizedLines = 
+      tokenizeLines
+        [| " $\"abc { { contents = 1 } }\"     " |]
```

Comment:
This doesn't seem to be a nested string interpolation. Am I misunderstanding the test name or it's an error?

Reply:
I think it's nested because there are two, non-doubled (as in: non-escaped) nested curlies. If it was only one, it wouldn't be nested. If the space between the curlies was removed, it wouldn't be interpolated.

Reply:
From the test name I was expecting nested string interpolations, like in `$"""{$""}"""`. The test source shows a normal interpolation (though an interesting case with with inner braces), and there's nothing I'd call nested. That's why I found it confusing.

Reply:
Ah, I see, yes indeed. Though "nested" can, imo, mean either: an interpolated string inside an interpolated string, or the expression part nested in itself: like a nested scope in C#, or nested lists etc.